actions.go

```diff
@@ -0,0 +1,123 @@
+type ExecuteAction func(s *discordgo.Session, m *discordgo.MessageCreate) error
+
+// ActionResolver is the interface used by Kadok to resolve a call into an action
+type ActionResolver interface {
+	GetPermission() security.Permission
+	ResolveAction(call []string) (ActionResolver, ExecuteAction)
+}
+
+// Action structure containing information and functions regarding an action
+type Action struct {
+	// The permission required to use this action. Use the EmptyPermission if no permission is required
+	Permission security.Permission
+	// The sub actions in the hierarchy. They do not inherit the permission requirement
+	SubActions map[string]ActionResolver
+	// The function to execute when the action is called
+	Execute func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error
+	// The function to display information about the action
+	DisplayHelp func(s *discordgo.Session, m *discordgo.MessageCreate) error
+}
+
+// MakeExecuteAction type from an Action
+func (action *Action) MakeExecuteAction(parameters []string) ExecuteAction {
+	if len(parameters) == 1 && strings.ToUpper(parameters[0]) == "HELP" {
+		return (*action).DisplayHelp
+	}
+	return func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+		return (*action).Execute(s, m, parameters)
+	}
+}
+
+// GetPermission is used with the security module to check the permission of the action
+func (action *Action) GetPermission() security.Permission {
+	return action.Permission
+}
+
+// ResolveAction looks if the calling parameters match a sub action it calls a ResolveAction on it, and return itself with an ExecuteAction if not.
+// If the parameter is help, return a DisplayHelp ExecuteAction
+func (action *Action) ResolveAction(call []string) (ActionResolver, ExecuteAction) {
+	if (len(call) == 0) || (len(call) == 1 && strings.ToUpper(call[0]) == "HELP") {
+		return action, action.MakeExecuteAction(call)
+	}
+	if subAction, ok := (*action).SubActions[strings.ToUpper(call[0])]; ok {
+		return subAction.ResolveAction(call[1:])
+	}
+	return action, action.MakeExecuteAction(call)
+}
+
+var (
+	// PingAction for Kadok to respond pong
+	PingAction = Action{
+		security.GetCharacterList,
+		map[string]ActionResolver{},
+		func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error {
+			_, err := s.ChannelMessageSend(m.ChannelID, "À Kadoc ! À Kadoc ! Pong!")
+			return err
+		},
+		func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+			return errors.New("Not a valid action")
+		},
+	}
+
+	// PongAction for Kadok to respond ping
+	PongAction = Action{
+		security.GetCharacterList,
+		map[string]ActionResolver{},
+		func(s *discordgo.Session, m *discordgo.MessageCreate, parameters []string) error {
+			_, err := s.ChannelMessageSend(m.ChannelID, "À Kadoc ! À Kadoc ! Ping!")
+			return err
+		},
+		func(s *discordgo.Session, m *discordgo.MessageCreate) error {
+			return errors.New("Not a valid action")
```

Comment:
where is that error handled ? it should display the help of the parent action, shouldn't it ?

Reply:
It is handled by the function `messageCreate` in `kadok.go`. It means that the help message is not yet implemented and should be implemented with #4 as `kadok help` action will require it to be present in some ways.